llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp

```diff
@@ -1314,8 +1314,8 @@ bool RISCVVLOptimizer::tryReduceVL(MachineInstr &OrigMI) {
     unsigned VLOpNum = RISCVII::getVLOpNum(MI.getDesc());
     MachineOperand &VLOp = MI.getOperand(VLOpNum);
 
-    if (!RISCV::isVLKnownLE(*CommonVL, VLOp)) {
-      LLVM_DEBUG(dbgs() << "    Abort due to CommonVL not <= VLOp.\n");
+    if (!RISCV::isVLKnownLT(*CommonVL, VLOp)) {
+      LLVM_DEBUG(dbgs() << "    Abort due to CommonVL not < VLOp.\n");
       continue;
     }
```

Comment:
Instead of defining isVLKnownLE, we could do

```suggestion
    if (!RISCV::isVLKnownLE(*CommonVL, VLOp)) {
      LLVM_DEBUG(dbgs() << "    Abort due to CommonVL not <= VLOp.\n");
      continue;
    }
    
    if (CommonVL->isIdenticalTo(VLOp)) {
      LLVM_DEBUG(dbgs() << "    Abort due to CommonVL == VLOp, no point in reducing.\n");
      continue
    }
```

That way we bail for correctness, and then bail to prune the worklist.

Reply:
Luke's suggestion allows an infinite loop if we somehow ended up with two VL values which were not identical, but represented equal VLs.  We can reduce to A, then reduce to B, then reduce to A, then B...  In principal, this could happen for two different register values which had the same runtime value.  I don't believe the current comparison logic makes this possible, but it would be a foot gun for the future.

I prefer @michaelmaitland 's original framing since it guarantees termination.  

Note that we could use the isIdenticalTo check in the implementation of the LT operator if we wanted.  This is more of an API design comment then an implementation one.

Reply:
> In principal, this could happen for two different register values which had the same runtime value.

If we have two different register values then `isVLKnownLE` would be false right? So we would bail anyway.